src/lang/eval/Eval.ml

```diff
@@ -378,6 +364,38 @@ let init_libraries clibs elibs =
             let%bind env = eres in
             init_lib_entry env lname lexp)
 
+(* Initializing libraries of a contract *)
+let init_libraries clibs elibs =
+  DebugMessage.plog ("Loading library types and functions.");
+  let%bind rec_env = init_lib_entries (pure Env.empty) Recursion.recursion_principles in
+  let rec recurser libnl =
+    if libnl = [] then pure rec_env else
+    (* Walk through library dependence tree. *)
+    foldM libnl ~init:[] ~f:(fun acc_env libnode ->
+      let dep_env = recurser libnode.deps in
+      let entries = libnode.libn.lentries in
+      let%bind env' = init_lib_entries dep_env entries in
+      (* Remove dep_env from env'. We don't want transitive imports.
```

Comment:
if `A` imports `B`, then `B` is loaded and "initialized" (which is essentially evaluated) and then with that environment, `A` is "initialized". Since `A` is all that is required (no transitive dependences), the environment of `A` alone is retained (the filtering). This filtering is done at each level.

Since `eval` is designed to add on to it's input environment, I don't see a non-messy, alternate way to do this.

Reply:
> Since `eval` is designed to add on to it's input environment

This is the key problem, I think. With non-transitive imports we are essentially saying that each import "level" is separate from the next, so the "correct" way of initialising a library is to build a new library separate from the ones that are being imported.

I'm happy with the filtering for now, but we should create an issue to fix `eval` for imports. It might be a good task for someone in the community to solve.

Reply:
>It might be a good task for someone in the community to solve.
#508